pkg/ingester/ingester.go

```diff
@@ -526,27 +526,14 @@ func (i *Ingester) starting(ctx context.Context) (err error) {
 	}
 
 	if i.ownedSeriesService != nil {
-		// We need to perform the initial computation of owned series after the TSDBs are opened but before the ingester becomes
-		// ACTIVE in the ring and starts to accept requests. However, because the ingester still uses the Lifecycler (rather
-		// than BasicLifecycler) there is no deterministic way to delay the ACTIVE state until we finish the calculations.
+		// Start owned series service asynchronously. We don't wait for ownedSeriesService to enter Running state here.
+		// This service will wait for entry in the ring, which is only created by lifecycler, and once the entry in the ring exists,
+		// ownedSeriesService will perform initial check of all tenants. While this check runs, we already allow read requests to proceed,
+		// but we block push requests (see checkAvailableForPushRequests).
 		//
-		// Since we don't actually need to be ACTIVE in the ring to calculate owned series (just present, with tokens) we instead
-		// calculate owned series before we start the lifecycler at all. Once we move the ingester to the BasicLifecycler and
-		// have better control over the ring states, we could perform the calculation while in JOINING state, and move to
-		// ACTIVE once we finish.
-		//
-		// TODO: When using partitions ring, we need to update all tenants after partition is registered into the ring. That is done by partitionLifecycler's starting method.
-
-		// Fetch and cache current ring state
-		_, err := i.ownedSeriesService.ringStrategy.checkRingForChanges()
-		switch {
-		case errors.Is(err, ring.ErrEmptyRing):
-			level.Warn(i.logger).Log("msg", "ingester ring is empty")
-		case err != nil:
-			return fmt.Errorf("can't read ring: %v", err)
-		default:
-			// We pass ringChanged=true, but all TSDBs at this point (after opening TSDBs, but before ingester switched to Running state) also have "new user" trigger set anyway.
-			i.ownedSeriesService.updateAllTenants(ctx, true)
+		// We pass ingester's service context to ownedSeriesService, to make ownedSeriesService stop when ingester exits Running state.
+		if err := i.ownedSeriesService.StartAsync(ctx); err != nil {
```

Comment:
That's not necessary, because we're passing ingester's service context to the `ownedSeriesService`, and `BasicService` implementation will [cancel this context when ingester's starting function returns with error.](https://github.com/grafana/dskit/blob/21930666b68cbeb57f71abad59211909c5aec5df/services/basic_service.go#L156-L170) That would in turn tell `ownedSeriesService` to stop.

Lifecycler is different in that we pass separate context to it.